sway-ir/src/block.rs

```diff
@@ -259,10 +259,26 @@ impl Block {
 #[doc(hidden)]
 impl BlockContent {
     pub(super) fn num_predecessors(&self, context: &Context) -> usize {
-        match &context.values[self.instructions[0].0].value {
-            ValueDatum::Instruction(Instruction::Phi(list)) => list.len(),
-            _ => unreachable!("First value in block instructions is not a phi."),
-        }
+        self.function.block_iter(context).fold(0, |mut acc, b| {
+            acc += b.instruction_iter(context).fold(0, |mut acc, i| {
+                acc += match context.values[i.0].value {
+                    ValueDatum::Instruction(Instruction::ConditionalBranch {
+                        true_block,
+                        false_block,
+                        ..
+                    }) => {
+                        true_block.get_label(context) == self.label
+                            || false_block.get_label(context) == self.label
+                    }
+                    ValueDatum::Instruction(Instruction::Branch(block)) => {
+                        block.get_label(context) == self.label
+                    }
+                    _ => false,
+                } as usize;
+                acc
+            });
+            acc
+        })
```

Comment:
I think it might be simpler to just iterate over all instructions in the function, filter for `br` and `cbr`, gather all the labels, filter for `self.label`, and take the length.  Not sure about perf complexity but it'd be easier to read.

Also, `fold` should very rarely have a mutable integer accumulator.  Instead you can just add to it.
```
iter().fold(0, |acc, i|) {
    acc + if blah { 0 } else { 1 }
})
```

Reply:
I pushed a fix for some merge conflicts below, and then since I was already on the branch, I  refactored the above:
```rust
        self.function
            .instruction_iter(context)
            .filter(|(_block, ins_value)| match &context.values[ins_value.0].value {
                ValueDatum::Instruction(Instruction::ConditionalBranch {
                    true_block,
                    false_block,
                    ..
                }) => {
                    true_block.get_label(context) == self.label
                        || false_block.get_label(context) == self.label
                }
                ValueDatum::Instruction(Instruction::Branch(block)) => {
                    block.get_label(context) == self.label
                }
                _otherwise => false,
            })
            .count()
```
It's using `Function::instruction_iter()` which is a little shortcut I made exactly for this sort of thing.  And then it takes your match expression and puts it in a filter and then just counts the number of matches with `Iterator::count()`.  A bit simpler and no mutable accumulators.

Reply:
Changed my mind.  That cast is just a little too janky. :slightly_smiling_face: Have pushed my change.

Reply:
Yup this is much better. I knew you'd hate that cast... just wanted a shortcut. Anyways, thank you for making the changes and for all the tips 😄